redis/src/main/java/org/apache/calcite/adapter/redis/RedisJedisManager.java

```diff
@@ -0,0 +1,115 @@
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+
+import org.slf4j.Logger;
+
+import java.util.Map;
+
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+import redis.clients.jedis.Protocol;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Manages connections to the Redis nodes.
+ */
+public class RedisJedisManager implements AutoCloseable {
+  private static final Logger LOGGER = CalciteTrace.getPlannerTracer();
+  private final LoadingCache<String, JedisPool> jedisPoolCache;
+  private final JedisPoolConfig jedisPoolConfig;
+
+  private final int maxTotal = GenericObjectPoolConfig.DEFAULT_MAX_TOTAL;
+  private final int maxIdle = GenericObjectPoolConfig.DEFAULT_MAX_IDLE;
+  private final int minIdle = GenericObjectPoolConfig.DEFAULT_MIN_IDLE;
+  private final int timeout = Protocol.DEFAULT_TIMEOUT;
+
+  private final String host;
+  private final String password;
+  private final int port;
+  private final int database;
+
+  public RedisJedisManager(String host, int port, int database, String password) {
+    JedisPoolConfig jedisPoolConfig = new JedisPoolConfig();
+    jedisPoolConfig.setMaxTotal(maxTotal);
+    jedisPoolConfig.setMaxIdle(maxIdle);
+    jedisPoolConfig.setMinIdle(minIdle);
+    this.host = host;
+    this.port = port;
+    this.database = database;
+    this.password = password;
+    this.jedisPoolConfig = jedisPoolConfig;
+    this.jedisPoolCache = CacheBuilder.newBuilder()
+        .removalListener(new JedisPoolRemovalListener())
+        .build(CacheLoader.from(this::createConsumer));
+  }
+
+  public JedisPool getJedisPool() {
+    requireNonNull(host, "host is null");
+    return jedisPoolCache.getUnchecked(host);
+  }
+
+  public Jedis getResource() {
+    return getJedisPool().getResource();
+  }
+
+  private JedisPool createConsumer() {
+    String pwd = password;
+    if (StringUtils.isEmpty(pwd)) {
+      pwd = null;
+    }
+    return new JedisPool(jedisPoolConfig, host, port, timeout, pwd, database);
+  }
+
+  /**
+   * JedisPoolRemovalListener for remove elements from cache.
+   */
+  private static class JedisPoolRemovalListener implements RemovalListener<String, JedisPool> {
+    @Override public void onRemoval(RemovalNotification<String, JedisPool> notification) {
+      assert notification.getValue() != null;
+      try {
+        notification.getValue().destroy();
+      } catch (Exception e) {
+        LOGGER.warn("While destroying JedisPool " + notification.getKey());
+      }
+    }
+  }
+
+  @Override public void close() {
+    for (Map.Entry<String, JedisPool> entry : jedisPoolCache.asMap().entrySet()) {
+      try {
+        entry.getValue().destroy();
+      } catch (Exception e) {
+        LOGGER.warn("While destroying JedisPool " + entry.getKey());
+      }
+    }
```

Comment:
Do we need L106~L112?

Reply:
> Do we need L106~L112?

I think it is necessary. There may be JedisPool in the cache when the cache is not removed by remove.

Reply:
> Do we need L106~L112?

Yes, you are right, I'm testing it now, L106~L112 does not need.